url73e3057d61_empeeoh_BACnet4J_tgz-pJ8-com_serotonin_bacnet4j_test_OverflowTestJ8/src/com/serotonin/bacnet4j/npdu/ip/WaitingRoom.java

```diff
@@ -0,0 +1,276 @@
+        Key key = new Key(peer, linkService, id, false);
+        Member member = new Member();
+
+        synchronized (waitHere) {
+            if (waitHere.get(key) != null)
+                throw new BACnetRuntimeException("Cannot enter a server into the waiting room. key=" + key);
+            waitHere.put(key, member);
+        }
+
+        return key;
+    }
+
+    public AckAPDU getAck(Key key, long timeout, boolean throwTimeout) throws BACnetException {
+        return (AckAPDU) getAPDU(key, timeout, throwTimeout);
+    }
+
+    public ConfirmedRequest getRequest(Key key, long timeout, boolean throwTimeout) throws BACnetException {
+        return (ConfirmedRequest) getAPDU(key, timeout, throwTimeout);
+    }
+
+    public Segmentable getSegmentable(Key key, long timeout, boolean throwTimeout) throws BACnetException {
+        APDU apdu = getAPDU(key, timeout, throwTimeout);
+        if (apdu instanceof Abort)
+            throw new SegmentedMessageAbortedException((Abort) apdu);
+        try {
+            return (Segmentable) apdu;
+        }
+        catch (ClassCastException e) {
+            throw new BACnetException("Receiving an APDU of type " + apdu.getClass()
+                    + " when expecting a Segmentable for key " + key);
+        }
+    }
+
+    public APDU getAPDU(Key key, long timeout, boolean throwTimeout) throws BACnetException {
+        Member member = getMember(key);
+        APDU apdu = member.getAPDU(timeout);
+        if (apdu == null && throwTimeout)
+            throw new BACnetTimeoutException("Timeout while waiting for APDU id " + key.getInvokeId());
+        return apdu;
+    }
+
+    public void leave(Key key) {
+        synchronized (waitHere) {
+            waitHere.remove(key);
+        }
+    }
+
+    public void notifyMember(InetSocketAddress peer, Address linkService, byte id, boolean isFromServer, APDU apdu)
+            throws BACnetException {
+        System.out.println("Received APDU (1): " + apdu);
+        if (apdu != null && apdu instanceof ComplexACK)
+            ((ComplexACK) apdu).parseServiceData();
+        System.out.println("Received APDU: " + apdu);
+        Key key = new Key(peer, linkService, id, isFromServer);
+        Member member = getMember(key);
+        if (member != null) {
+            member.setAPDU(apdu);
+            return;
+        }
+
+        // The member may not have gotten around to listening for a message yet, so enter a retry loop to
+        // make sure that this message gets to where it's supposed to go if there is somewhere to go.
+        int attempts = 5;
+        long sleep = 50;
+        while (attempts > 0) {
+            member = getMember(key);
+            if (member != null) {
+                member.setAPDU(apdu);
+                return;
+            }
+
+            attempts--;
+            try {
+                Thread.sleep(sleep);
+            }
+            catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)